packages/devextreme/js/ui/date_box/ui.date_box.mask.js

```diff
@@ -304,7 +304,14 @@ const DateBoxMask = DateBoxBase.inherit({
     },
 
     _searchString(char) {
-        if(!isNaN(parseInt(this._getActivePartProp('text')))) {
+        const activePartText = this._getActivePartProp('text');
+
+        if(!isNaN(parseInt(activePartText))) {
+            return;
+        }
+
+        if((char.toLowerCase() === 'p' && activePartText === 'AM') || (char.toLowerCase() === 'a' && activePartText === 'PM')) {
```

Comment:
It's not a good approach to add specific checks for each found scenario:
- it's difficult to consider everything. In the note below I already found one additional scenario => this `if` would grow much;
- it's difficult to maintain. If we wanna extend the behavior in future we will need to rewrite attentively this conditions;
- if we add specific conditions we don't improve a general algorithm => neighbor scenarios still can be broken.

Did u find why general algorithm does not work? Seems like everything is logical: we brut force all values for active part and check if it contains the passed search string. 

I prefer we debug this algorithm more, find the exact reason and try to improve it.